### PR TITLE
[AUG] Add MNS and Correct sampling methods

### PR DESCRIPTION
**Pull request by Tokkiu**
* MNS https://research.google/pubs/pub50257
* Correct https://research.google/pubs/pub48840

### CODE REVIEW

File: model_zoo/CorrectDNN/src/Correct.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+
+        # Update array A and B
+        self.step += 1
+        for item in neg_in_unis:
+            self.arr_B[item] = (1-self.lr)*self.arr_B[item] + self.lr*(self.step-self.arr_A[item])
+            self.arr_A[item] = self.step
+
+        return torch.tensor(sampled_items).to(samples.device)
+
+    def forward(self, inputs):
```
Comment (aporia3517) [inline model_zoo/CorrectDNN/src/Correct.py:86]: @Tokkiu 
My guess is that if we only use in-batch negatives, we don't need to do any sampling. If we implement the forward statement as follows, we can implement a sampled softmax loss using in-batch negatives. (please refer `else` clause.)

```
    def forward(self, inputs):
        """
        Inputs: [user_dict, item_dict, label]
        """
        user_dict, item_dict, labels = inputs[0:3]
        user_vecs = self.user_tower(user_dict)
        # print('forward user_vecs', user_vecs)
        user_vecs = self.user_dropout(user_vecs)
        item_vecs = self.item_tower(item_dict)
        # print('forward item_vecs', item_vecs)
        item_vecs = self.item_dropout(item_vecs)
        if not self.in_batch_negatives:
            y_pred = torch.bmm(
                item_vecs.view(-1, self.num_negs + 1, self.output_dim),
                user_vecs.unsqueeze(-1),
            ).squeeze(-1)
            loss = self.get_total_loss(y_pred, labels)
            return_dict = {"loss": loss, "y_pred": y_pred}
        else:
            # item_vecs: [B x (1+num_negs) x output_dim]
            # user_vecs: [B x output_dim]
            batch_items = item_dict["item_id"].view(
                user_vecs.size(0), self.num_negs + 1
            )
            pos_items = batch_items[:, 0].cpu().numpy()
            in_batch_freqs = Counter(pos_items) # Dict of {item_id: freq} # len(in_bathc_freqs) = batch_size
            logQ_correction = (
                torch.log(
                    torch.tensor(
                        [in_batch_freqs[i] for i in pos_items], dtype=torch.float32
                    )
                    + 1e-6
                )
                .reshape((1, -1))
                .to(self.device)
            )
            viewed_item_vecs = item_vecs.view(-1, self.num_negs + 1, self.output_dim)
            pos_item_vecs = viewed_item_vecs[:, 0, :] # positive item만 slicing
            score = self.cos(user_vecs.unsqueeze(1), pos_item_vecs.unsqueeze(0)) # B x B
            score -= logQ_correction
            labels = torch.arange(score.size(0)).long().to(self.device)

            loss = self.loss_fn(score, labels)  # need to add regularizaiton loss
            return_dict = {"loss": loss}

        return return_dict
```

I'll share the model_config.yaml file and full src code on slack for reproduction.
